### ISSUE TITLE
fix(lints): Warn on unwrap/expect/panic in library crates via workspace lints

### ISSUE THREAD

**Issue by d-o-hub** [d-o-hub/rust-self-learning-memory#800]
## Problem

`unwrap()`, `expect()`, and `panic!()` in library code are silent correctness risks. When set to `"allow"` globally, new code introduces panics without any warning. CI should catch these at the lint level.

### 1. Add `[workspace.lints.clippy]` to root `Cargo.toml`

Move lint configuration to workspace level so all crates inherit it:

```toml
[workspace.lints.clippy]
# Correctness: warn on panic-inducing patterns in library code
unwrap_used = "warn"
expect_used = "warn"
panic = "warn"

# Enable all clippy lints at warning level
all = { level = "warn", priority = -1 }
pedantic = { level = "allow", priority = -1 }
nursery = { level = "allow", priority = -1 }

# Promoted pedantic lints for correctness
cast_precision_loss = "warn"
cast_possible_truncation = "warn"
map_unwrap_or = "warn"

[workspace.lints.rust]
unsafe_code = "allow"  # Only if SIMD/FFI is used
```

### 2. Inherit in each sub-crate `Cargo.toml`

```toml
[lints]
workspace = true
```

### 3. Override in benchmarks/examples/CLI crates

```toml
[lints]
workspace = true

[lints.clippy]
unwrap_used = "allow"
expect_used = "allow"
panic = "allow"
```

### 4. Add `#![allow(...)]` in test modules and test files

For every `#[cfg(test)] mod tests { }` block:
```rust
#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
    // ... test code
}
```

For standalone test files (`tests/*.rs`, `benches/*.rs`, `examples/*.rs`):
```rust
#![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
// ... file content
```

### 5. Add targeted `#[allow]` for justified production usage

For lock poisoning, metric initialization, and other justified panics:
```rust
#[allow(clippy::expect_used)] // Lock poisoning is unrecoverable
let cache = self.cache.read().expect("cache lock poisoned");
```

## Acceptance Criteria

- [ ] Workspace-level lint set to `"warn"` for `unwrap_used`, `expect_used`, `panic`
- [ ] All sub-crates use `[lints] workspace = true`
- [ ] Benchmarks/examples/CLI have per-crate overrides to `"allow"`
- [ ] All `#[cfg(test)]` modules have `#![allow(...)]`
- [ ] All `tests/*.rs` and `examples/*.rs` files have file-level `#![allow(...)]`
- [ ] `cargo clippy --all-targets -- -D warnings` passes clean
- [ ] No new unwrap/expect usage triggers warnings in library code

## Impact

- **New code**: Gets immediate feedback if `unwrap()`/`expect()` is used in library paths
- **Existing code**: Grandfathered with targeted allows; can be cleaned incrementally
- **Tests**: Unaffected (blanket allows)
- **CI**: `-D warnings` in RUSTFLAGS will catch new violations

## References

- Implemented in `d-o-hub/chaotic_semantic_memory` PR #497
- DeepSource Rust analyzer parity
- Clippy docs: https://rust-lang.github.io/rust-clippy/
